### PR TITLE
Calculate the time of execution in the hierarchy

### PR DESCRIPTION
**Pull request by arek1-618**
Calculate the time of execution in the hierarchy

### CODE REVIEW

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Arkadiusz Adach <a.adach@yetiforce.com>
+ */
+
+/**
+ * CalculateSumOfExecutionTime class.
+ */
+class OSSTimeControl_CalculateSumOfExecutionTime_Model
```
Comment (rskrzypczak) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]: shorter names of variables and methods

File: install/install_schema/scheme.sql [hunk 2]
```diff
@@ -8703,6 +8703,7 @@ CREATE TABLE `vtiger_troubletickets` (
   `response_expected` datetime DEFAULT NULL,
   `solution_expected` datetime DEFAULT NULL,
   `idle_expected` datetime DEFAULT NULL,
+  `total_execution_time` int(10) DEFAULT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: total_execution_time > sum_time_subordinate

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php (hunk 1 above)
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]:
```suggestion
class OSSTimeControl_TimeCounting_Model
```

File: modules/com_vtiger_workflow/tasks/VTUpdateWorkTime.php [hunk 3]
```diff
@@ -47,8 +47,8 @@ public function doTask($recordModel)
 		$metasData = vtlib\Functions::getCRMRecordMetadata(array_keys($referenceIds));
 		$modulesHierarchy = array_keys(App\ModuleHierarchy::getModulesHierarchy());
 		foreach ($metasData as $referenceId => $metaData) {
-			if (((int) $metaData['deleted']) === 0 && in_array($metaData['setype'], $modulesHierarchy)) {
-				OSSTimeControl_Record_Model::recalculateTimeControl($referenceId, $referenceIds[$referenceId]);
+			if (0 === ((int) $metaData['deleted']) && \in_array($metaData['setype'], $modulesHierarchy)) {
+				\OSSTimeControl_CalculateSumOfExecutionTime_Model::calculateAndUpdate($referenceId, $referenceIds[$referenceId]);
```
Comment (mariuszkrzaczkowski) [inline modules/com_vtiger_workflow/tasks/VTUpdateWorkTime.php]:
```suggestion
				\OSSTimeControl_CalculateSumOfExecutionTime_Model::calculateAndUpdate($referenceId, $metaData['setype'],  $referenceIds[$referenceId]);
```

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+	/**
+	 * Calculate and update.
+	 *
+	 * @param int    $recordId
+	 * @param string $name
+	 *
+	 * @return void
+	 */
+	public static function calculateAndUpdate(int $recordId, string $name)
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]:
```suggestion
	public static function calculateAndUpdate(int $recordId, string $moduleName, string $relationField)
```

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+	/**
+	 * Recalculate time control.
+	 *
+	 * @param int    $recordId
+	 * @param string $name
+	 *
+	 * @return void
+	 */
+	public function recalculateTimeControl(int $recordId, string $name)
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]:
```suggestion
	public function recalculateTimeControl(int $recordId, string $relationField)
```

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	/**
+	 * Get sum of time.
+	 *
+	 * @param int    $recordId
+	 * @param string $name
+	 *
+	 * @return float
+	 */
+	private function getSumTime(int $recordId, string $name): float
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]:
```suggestion
	private function getSumTime(int $recordId, string $relationField): float
```

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+	/**
+	 * Construct.
+	 *
+	 * @param string $moduleName
+	 */
+	public function __construct(string $moduleName)
+	{
+		$this->moduleName = $moduleName;
+		$this->fieldNameSumOfTime = 'sum_time';
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]: sum_time this is a constant

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+	 *
+	 * @param int    $recordId
+	 * @param string $name
+	 *
+	 * @return void
+	 */
+	public static function calculateAndUpdate(int $recordId, string $name)
+	{
+		$moduleName = 'HelpDesk';
+		$instance = new self($moduleName);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]:
```suggestion
		$instance = new self($moduleName, $recordId, $relationField);
```

File: modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	 * @return void
+	 */
+	private function update(int $recordId, float $sumTime)
+	{
+		\App\Db::getInstance()
+			->createCommand()
+			->update(
+				$this->tableName,
+			[$this->columnName => $sumTime],
+			['ticketid' => $recordId]
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/CalculateSumOfExecutionTime.php]: ?

File: modules/OSSTimeControl/models/TimeCounting.php [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+	 * @var \Vtiger_Module_Model
+	 */
+	private $moduleModel;
+
+	/**
+	 * Construct.
+	 *
+	 * @param string $moduleName
+	 */
+	public function __construct(string $moduleName)
```
Comment (mariuszkrzaczkowski) [inline modules/OSSTimeControl/models/TimeCounting.php]:
```suggestion
	public function __construct(string $moduleName, int $recordId, string $relationField)
```

**Pull request merged**